### PR TITLE
Add constant expr

### LINKED ISSUE #18385: Minimal support for Const expression
**Description:**
This requires to support following in float lang-lib
```
# IEEE not-a-number value
public const float NaN = 0.0/0.0;

# IEEE positive infinity
public const float Infinity = 1.0/0.0;
```
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by hasithaa**
Fixes #18385.
Fixed https://github.com/ballerina-platform/ballerina-lang/issues/18636

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 1]
```diff
@@ -777,8 +777,11 @@ expression
     ;
 
 constantExpression
-    :   simpleLiteral
-    |   recordLiteral
+    :   simpleLiteral                                                       # constSimpleLiteralExpression
+    |   recordLiteral                                                       # constRecordLiteralExpression
+    |   constantExpression (DIV | MUL) constantExpression                   # constDivMulModExpression
+    |   constantExpression (ADD | SUB) constantExpression                   # constAddSubExpression
+    |   LEFT_PARENTHESIS constantExpression RIGHT_PARENTHESIS               # constGroupExpressions
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]:
```suggestion
    |   LEFT_PARENTHESIS constantExpression RIGHT_PARENTHESIS               # constGroupExpression
```

Should this be `constGroupExpression` instead?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -4234,7 +4234,18 @@ public void visit(BLangJSONArrayLiteral jsonArrayLiteral) {
 
     @Override
     public void visit(BLangConstant constant) {
-        constant.expr = rewriteExpr(constant.expr);
+
+        BConstantSymbol constSymbol = constant.symbol;
+        if (constSymbol.literalType.tag <= TypeTags.BOOLEAN || constSymbol.literalType.tag == TypeTags.NIL) {
+            BLangLiteral literal = ASTBuilderUtil.createLiteral(constant.expr.pos, constSymbol.literalType,
+                    constSymbol.value.value);
+            if (constSymbol.literalType.tag != TypeTags.NIL && constSymbol.value.value == null) {
```
[the remark this answers is not in the document]
Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: done

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 3]
```diff
@@ -126,17 +135,182 @@ public void visit(BLangRecordLiteral recorLiteral) {
...
+        } catch (ArithmeticException ae) {
+            dlog.error(currentPos, DiagnosticCode.INVALID_CONST_EXPRESSION, ae.getMessage());
+        }
+        // This is a compilation error already logged.
+        // This is to avoid NPE exceptions in sub-sequent validations.
+        return new BLangConstantValue(null, this.currentConstSymbol.type);
+    }
+
+    private BLangConstantValue calculateAddition(BLangConstantValue lhs, BLangConstantValue rhs) {
+        // TODO : Handle overflow in numeric values.
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java]: Shall we create an issue for this?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 4]
```diff
@@ -1138,6 +1135,12 @@ error.cannot.define.constant.with.type=\
 error.expression.is.not.a.constant.expression=\
   expression is not a constant expression
 
+error.invalid.const.expression=\
+  invalid constant expression, reason ''{0}''
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: Minor change, how about something like
```suggestion
  invalid constant expression: {0}
```
instead?

**Pull request merged**